LFS-238: Integrate the pedigree editor

allasm opened the pull request.

allasm committed:
LFS-238: Integrate the pedigree editor

New pedigree module
- includes initial package.json with all the required packages

allasm committed:
LFS-238: Integrate the pedigree editor

Initial port of the OpenPedigree version of the pedigree editor.

- all pedigree files are added under AGPL v3 license

- the port includes a big prototype.js shim, since prototype.js is completely incompatible with React;
  only a subset of prototype.js used in pedigree is ported; some functions are rewritten from scratch,
  others are copied. Committed under MIT license, same as prototype.js

- similarly a PhenoTips widget used in the editor had to be ported since it also relied on prototype.js

- font-awesome used by the editor was rolled back to version 4 (compared to v5 that open pedigree uses)
  since v4 uses css and v5 uses svg, which breaks all events (open pedigree is broken);
  css is included in this commit (under its own license), the font used by fa css is loaded from the web

- this port is fully functional, except a few pieces that require more porting:
  - gene\phenotype\diagnosis suggestions
  - export to a file
  - zoom slider and abnormality dragging (the library used by both is extremely prototype.js dependant)

allasm committed:
LFS-238: Integrate the pedigree editor

Added AGPL to the list of licenses recognized by rat plugin
Excluded some included 3rd-party files (raphael.js, prototype.js shim, font-awesome css) from license checks

allasm committed:
LFS-238: Integrate the pedigree editor

Enable CSS loading from .js files in aggregated frontend (in webpack.config-template.js)

allasm committed:
LFS-238: Integrate the pedigree editor

Enable pedigree question in the Demographics section

allasm committed:
LFS-238: Integrate the pedigree editor

Initial implementation of PedigreeQuestion that opens the editor when the pedigree image is clicked
 - the editor works in a full-screen modal, can be opened and closed
 - pedigree is saved to LFS
 - placeholder image and its size are hardcoded

allasm commented:
Addressed codereview comments:
- pedigree is in a separate module
- all files have license headers. A couple of 3-rd party libraries are excluded from license checks

allasm committed:
LFS-238: Integrate the pedigree editor

bugfix: added an empty folder required for the build process that git did not add automatically

allasm committed:
LFS-238: Integrate the pedigree editor

Pedigree editor bugfixes:
 - restore original window.onbeforeunload when pedigree is closed
 - fix bug where legends refer to incorrect nodes after internal node IDs are changed

allasm committed:
LFS-238: Integrate the pedigree editor

Pedigree UI fixes:
 - disabled non-functional zoom slider
 - fixed templates which are sometimes displayed incorrectly (simplified fill from gradient to plain)

allasm committed:
LFS-238: Integrate the pedigree editor

Implemented file export with no external code (which was only needed for compatibility with old browsers)

veronikaslc commented on modules/pedigree/src/main/frontend/package.json:
```diff
@@ -0,0 +1,51 @@
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/ccmbioinfo/lfs.git",
+    "directory": "modules/data-entry/src/main/frontend"
```
Wrong directory

veronikaslc commented:
In the `Personal` tab any change erases dates. 
![image](https://user-images.githubusercontent.com/4371097/80157809-fd53c100-8594-11ea-97af-1f54fb3cbb69.png)

Same for `Clinical` tab and `comments`.

veronikaslc commented:
Typing in `comments` gives error:
![image](https://user-images.githubusercontent.com/4371097/80157915-41df5c80-8595-11ea-89f6-f53a357594fc.png)

veronikaslc commented:
Error after saving and closing pedigree:
![image](https://user-images.githubusercontent.com/4371097/80158064-92ef5080-8595-11ea-9f52-3a4a81dfa764.png)

sashaandjic commented:
My understanding of the task was to integrate pedigree into the Demographic form with no functionality.
I was able to create a family tree, assign names to the nodes, gender and adopted status. When I saved and closed the pedigree, the correct pedigree was shown in the form. However, upon refreshing it was obvious that the pedigree was not saved.

As per my understanding and previous testing, this is the minimum we wanted to achieve in this task.

allasm committed:
LFS-238: Integrate the pedigree editor

Addressed codereview comments: fixed package.json

allasm committed:
LFS-238: Integrate the pedigree editor

Addressed codereview comments: fixed bug in pedigree cleanup and in comment handling in node menu

allasm committed:
LFS-238: Integrate the pedigree editor

Fixed bugs in Prototype.js shim and added Ajax implementation to the shim

allasm committed:
LFS-238: Integrate the pedigree editor

Added Phenotips suggest widget shim and related CSS
(and a small xwiki shim to substitute XWiki functions used by the widget)

allasm committed:
LFS-238: Integrate the pedigree editor

Fixed some pedigree files to use the new Ajax shim

allasm committed:
LFS-238: Integrate the pedigree editor

Enabled suggest picker for disorders, genes and phenotypes (with blank URL for now)

sashaandjic commented:
Most of the pedigree functionality is present and working properly. Here are some issues that I found:
- DOB and DOD do not get saved; also there is no date format validation
- saving a pedigree and closing it does not save the pedigree changes; the save happens only if I hit SAVE bttn on the form; however if I initially saved the form and then opened pedigree, change, save and close it, the form will have SAVED bttn only. After refresh all of my pedigree changes will be gone
- export does not take into account changes that were not saved (e.g. I added some nodes without saving and exported pedigree; the export file did not have new nodes)
- export dialog's **(Select a local file to be imported)** does not look like a button (although behaves as one)
![image](https://user-images.githubusercontent.com/23243244/81645522-b91b4a00-93f7-11ea-87d6-ac582ea2a9fa.png)

All in all it looks good. All reported problems are not show stoppers.

veronikaslc commented:
Confirm all of above errors, plus:
- Any changes made in personal tab still erases DOB and DOD
- All requests for term lookup for disease, genes and phenotips inputs in the clinical tab go to `content.html/Forms/q=xx`

Commit:
LFS-238: Integrate the pedigree editor
Fixed DOB and DOD functioning

veronikaslc commented:
Closed in favour of https://github.com/ccmbioinfo/lfs/pull/293
